### PR TITLE
UI for preference rollout

### LINKED ISSUE #63: UI for preference rollout
Copied from: https://github.com/mozilla/normandy/issues/1197

> We'll need a UI to create preference rollout recipes. Note that this is going to be implemented as a "local action" in the normandy client. See [bug 1440782](https://bugzilla.mozilla.org/show_bug.cgi?id=1440782) for more info.
> 
> This will likely have to wait until the local action is speced out so we know what arguments will be needed
> 
> ### [Preference roll-out blueprint](https://docs.google.com/document/d/1qFCOscDiMGVDcBYtznzgPNCWzOhaxC4X5VGt3LTjZY0/edit#heading=h.eoqr3si0tn8z)

### PR DESCRIPTION
**Pull request by peterbe**
Fixes #63 

- [x] Unit tests of the component beyond just `expect(wrapper).not.toThrow();`
- [x] Use snapshots in unit tests
- [x] Figure out if the "Add row" button included in the form submission matters
- [x] Figure out if the `data.arguments.preferences` should be displayed differently on the Recipe Details page 
- [x] Test deleting rows and saving. Should not 400 error.

### CODE REVIEW

**Comment by peterbe** [comment]
@rehandalal Before asking for a review officially, can you take a quick look at the tests and see if it's enough? I never figured out how to not-shallow render the components (`PreferenceRolloutFields` or `RowField`) because of weird errors related to `<FormItem>`. If you have some tips on how to do non-shallow rendering in tests with components that use FormItem, I'm all ears.

File: src/tests/workflows/recipes/components/PreferenceRolloutFields.test.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    // it('should be possible to remove a row', () => {
+    //   const props = {
+    //     row: {
+    //       name: 'name',
+    //       type: 'string',
+    //       value: 'Peter',
+    //     },
+    //   };
+
+    //   // const wrapper = mount(<RowField {...props} />);
```
Comment (peterbe) [inline src/tests/workflows/recipes/components/PreferenceRolloutFields.test.js]: Here's a terrible start to how I tried to render the `RowField` component. My goal was to simulate input and simulate clicks on buttons and stuff. Could do with some help on how to mock `FormItem`.

File: src/workflows/recipes/components/RecipeForm.js [hunk 2]
```diff
@@ -27,6 +30,11 @@ export function cleanRecipeData(data) {
       error.data = { arguments: 'Invalid JSON.' };
       throw error;
     }
+  } else if (data.arguments.preferences) {
+    // XXX How is this going to work when it's PreferenceRollback??
```
Comment (peterbe) [inline src/workflows/recipes/components/RecipeForm.js]: Not sure what to do about this comment.

File: src/workflows/recipes/components/PreferenceRolloutFields.js [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+              label="Slug"
+              name="arguments.slug"
+              initialValue={recipeArguments.get('slug', '')}
+            >
+              <Input disabled={disabled} />
+            </FormItem>
+          </Col>
+        </Row>
+
+        <p className="action-info">Preferences</p>
```
Comment (peterbe) [inline src/workflows/recipes/components/PreferenceRolloutFields.js]: This doesn't look great. I can't see any other similar examples of headers for a group of inputs.

File: src/workflows/recipes/components/RecipeForm.js (hunk 2 above)
Comment (rehandalal) [reply src/workflows/recipes/components/RecipeForm.js]: preference rollback only takes `rolloutSlug` as an argument, which is the slug of the preference rollout so it would never have `data.arguments.preferences` and we would not need to worry about this.

**Comment by peterbe** [comment]
By the way, I think this new feature is the first to use jest snapshots. I'm having serious doubts about that. In fact, I'm having serious doubts about using enzyme and shallow rendering at all. 

We can chat about that. Perhaps the (new) testing strategies I introduce in this PR do suck but perhaps we shouldn't let that block the release of this feature.

File: src/workflows/recipes/components/PreferenceRolloutFields.js [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        </Button>
+      </div>
+    );
+  }
+}
+
+export default PreferenceRolloutFields;
+
+export class RowField extends React.PureComponent {
+  state = {
```
[the remark this answers is not in the document]
Comment (mythmon) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: More information about the pitfalls of seeding state with props: https://medium.com/@justintulk/react-anti-patterns-props-in-initial-state-28687846cc2e

File: src/workflows/recipes/components/RecipeForm.js (hunk 2 above)
Comment (rehandalal) [reply src/workflows/recipes/components/RecipeForm.js]: we can probably just ditch this comment.

File: src/workflows/recipes/components/PreferenceRolloutFields.js (hunk 3 above)
Comment (rehandalal) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: yeah i'll do a quick style/css pass over this once it lands.

File: src/workflows/recipes/components/PreferenceRolloutFields.js [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    const { disabled, row, disableRemove } = this.props;
+    const { id } = row;
+    const { name, type, value } = this.state;
+
+    let valueInput;
+    if (type === 'string') {
+      valueInput = (
+        <Input
+          disabled={disabled}
+          onChange={event => this.handleValueChange(event.target.value)}
```
[the remark this answers is not in the document]
Comment (rehandalal) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: +1 for event handlers using the event object rather than things derived from it.

File: src/workflows/recipes/components/PreferenceRolloutFields.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterbe) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: I think I made it like that because it was the very first thing I did and I thought I would want to do more with these values.

File: src/workflows/recipes/components/PreferenceRolloutFields.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterbe) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: > is bad for performance

Not necessarily. 
See https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578
and https://twitter.com/ryanflorence/status/905574498400845825?lang=en and https://twitter.com/ryanflorence/status/911023769765871616

I *like* having inline functions because they're local. Meaning, their functionality is there right near where it matters and gets used. If you use public field methods or bound functions you have to scroll away from where it's used. 

Sometimes it's nice to break out into bound functions (or better, public field methods) to avoid DRY. E.g.

```javascript
class C extends React.Component {
  reload = () => { window.location.reload(true) }
  render() {
    return <div>
      <button onClick={this.reload}>Reload</button>
      ...
      <button onClick={this.reload}>Reload</button>
    </div>
  } 
}
```
is arguably better than:
```javascript
class C extends React.Component {
  render() {
    return <div>
      <button onClick={() => { window.location.reload(true) }}>Reload</button>
      ...
      <button onClick={() => { window.location.reload(true) }}>Reload</button>
    </div>
  } 
}
```

Also, the [Input](https://github.com/ant-design/ant-design/blob/32c78b2f8dd7f3b3ec059e956f07f083417259de/components/input/Input.tsx#L27) component is based on `React.Component` without a `shouldComponentUpdate` so it doesn't even bother doing a diff. It just goes straight to render. React's virtual DOM will concern itself whether to update the DOM or not and that's usually the genuinely expensive thing. 

As a matter of fact; we should probably reconsider our heavy usage of `React.PureComponent` anyway. As Ryan pointed out in his blog post, he benchmarked his app before he added `PureRenderMixin` (what it was before `React.PureComponent` was part of the core) and after. 
> "Much to my surprise, my app got slower 🤔."

That's because the majority of the time, the props (or state) were different and most of the time needed to re-render anyway so the shallow diffs that `React.PureComponent` had to do were just costing time. 

The real point of Ryan's blog post is that "premature optimization is evil" and that real optimization is measurements. There are probably enormously larger fish to fry in areas of the code that gets used a heck lot more. 

**In conclusion...** 

* Just because a `render` function executes doesn't mean it's bad for performance. What matters is updates to the DOM.
* Doing diffs like `React.PureComponent` does, might cost more than it saves in avoidances of calling `render`.
* I prefer the style of inline functions because it makes the code nicer to read often. I prefer to refactor that out to public field methods to avoid DRY or if feel there's some other reason to do so for clarity.
* Performance isn't about gut feelings but about measurements.
* Even if there's a small performance difference, in either direction, this deep in the component hierarchy, it really doesn't matter that much.

File: src/workflows/recipes/components/PreferenceRolloutFields.js [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        />
+      );
+    } else if (type === 'boolean') {
+      valueInput = <SwitchBox disabled={disabled} onChange={this.handleValueChange} />;
+    } else if (type === 'integer') {
+      valueInput = <InputNumber disabled={disabled} onChange={this.handleValueChange} />;
```
[the remark this answers is not in the document]
Comment (peterbe) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: They *are* controlled inputs. THe initial value is passed to the wrapping `FormItem` component. The `onChange` is for every character input. No?

[the remark this answers is not in the document]
Comment (mythmon) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: `FormItem` is passed an `initialValue`, which is the common sign of an uncontrolled component. However, the form handling here is complex enough that I'm not sure either way anymore.

File: src/workflows/recipes/components/PreferenceRolloutFields.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterbe) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: Now I remember why I did it the way I did it. 
I wanted to make my public field methods disconnected from the minutea of inputs and event handling stuff. The field methods `handleValueChange`, `handleNameChange`, and `handleTypeChange` gets the real data. Not events. 

The `onChange` prop on components like `SwitchBox` and `InputNumber` (both from `antd`) don't work like your regular pure React `<input type="text" value={somevalue} onChange={eventhandler}/>`. What you get back when you set an `onChange` prop on `SwitchBox` or `InputNumber` is the "cleaned up" value. Not an `event` object. **The `Input` component is different.** There you have to "convert" the `event` object to get to the actual input value. 

Had I instead opted for something like this:
```javascript
<Input onChange={this.handleNameChange}/>
```
then, the `handleNameChange` method would have to be different from all the other `handle*Change` methods there. I.e.:

```javascript
handleNameChange = event => {
  const name = event.target.value;
  const { id, type, value } = this.props;
  this.props.onRowChange({
    id,
    name,
    type,
    value,
  });
};
```

So there is consistency in the handler methods but some inconsistency in the `onChange` props usage.

[the remark this answers is not in the document]
Comment (rehandalal) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: To me, this looks clearer than what we have now and I think we should go with it:
```js
handleNameChange = event => {
  const name = event.target.value;
  const { id, type, value } = this.props;
  this.props.onRowChange({
    id,
    name,
    type,
    value,
  });
};
```

[the remark this answers is not in the document]
Comment (peterbe) [reply src/workflows/recipes/components/PreferenceRolloutFields.js]: Thing is, if you do this, all the handlers that were quite "pure" before are now mixed. Some take an `event`, some take an actual value from the controlled antd component. See [this list of methods](https://github.com/mozilla/delivery-console/blob/f181fd8050b540ff22468a1e2f74616a7672a3e7/src/workflows/recipes/components/PreferenceRolloutFields.js#L192-L230) now. More mixed bag. 

But I don't care enough. I removed all the discussed inline functions so it's `<Input onChange={this.handlerFunction}/>` etc.

**Comment by rehandalal** [comment]
bors retry

**Comment by mythmon** [comment]
Bors failed both times with this error:

```
{{:badmatch, {:error, :push, 422}},
 [{BorsNG.GitHub, :push!, 3,
   [file: 'lib/github/github.ex', line: 64]},
  {BorsNG.Worker.Batcher, :complete_batch, 3,
   [file: 'lib/worker/batcher.ex', line: 381]},
  {BorsNG.Worker.Batcher, :maybe_complete_batch, 1,
   [file: 'lib/worker/batcher.ex', line: 368]},
  {BorsNG.Worker.Batcher, :handle_cast, 2,
   [file: 'lib/worker/batcher.ex', line: 77]},
  {:gen_server, :try_dispatch, 4,
   [file: 'gen_server.erl', line: 616]},
  {:gen_server, :handle_msg, 6,
   [file: 'gen_server.erl', line: 686]},
  {:proc_lib, :init_p_do_apply, 3,
   [file: 'proc_lib.erl', line: 247]}]}
```

I *think* that means that Github rejected the push for some reason, but I'm not sure why.

**Comment by rehandalal** [comment]
third time's the charm?

bors retry
